File: lib/commons/src/main/java/io/github/jinganix/webpb/commons/SegmentGroup.java
/*
 * Copyright (c) 2020 The Webpb Authors, All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * https://github.com/jinganix/webpb
 */

package io.github.jinganix.webpb.commons;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Group of {@link UrlSegment} captured from url. */
public class SegmentGroup {

  private static final String KEY = "key";

  private static final String VALUE = "value";

  private static final Pattern PATH_PATTERN = Pattern.compile("(?<value>\\{[^/]+})");

  private static final Pattern QUERY_PATTERN =
      Pattern.compile("((?<key>\\w+)=)?(?<value>[^/?&]+)&?");

  private final List<UrlSegment> pathSegments = new ArrayList<>();

  private String suffix = "";

  private final List<UrlSegment> querySegments = new ArrayList<>();

  private final List<UrlSegment> segments = new ArrayList<>();

  /** Constructor. */
  public SegmentGroup() {}

  /**
   * Static creator.
   *
   * @param url request url
   * @return Params group
   */
  public static SegmentGroup of(String url) {
    SegmentGroup group = new SegmentGroup();
    if (url == null || url.isEmpty()) {
      return group;
    }
    String[] parts = url.split("\\?");
    String path = parts[0];
    String query = null;
    if (parts.length == 2) {
      query = parts[1];
    } else if (path.contains("=")) {
      path = null;
      query = parts[0];
    }

    if (path != null) {
      Matcher pathMatcher = PATH_PATTERN.matcher(path);
      int index = 0;
      while (pathMatcher.find()) {
        UrlSegment segment =
            new UrlSegment(
                path.substring(index, pathMatcher.start()), null, pathMatcher.group(VALUE));
        group.pathSegments.add(segment);
        index = pathMatcher.end();
      }
      group.suffix = path.substring(index);
    }

    if (query != null) {
      Matcher queryMatcher = QUERY_PATTERN.matcher(query);
      while (queryMatcher.find()) {
        UrlSegment segment =
            new UrlSegment(null, queryMatcher.group(KEY), queryMatcher.group(VALUE));
        group.querySegments.add(segment);
      }
    }
    group.segments.addAll(group.pathSegments);
    group.segments.addAll(group.querySegments);
    return group;
  }

  /**
   * If params is empty.
   *
   * @return true if params is empty
   */
  public boolean isEmpty() {
    return pathSegments.isEmpty() && querySegments.isEmpty();
  }

  /**
   * Get path segments.
   *
   * @return list of {@link UrlSegment}
   */
  public List<UrlSegment> getPathSegments() {
    return pathSegments;
  }

  /**
   * Get query segments.
   *
   * @return list of {@link UrlSegment}
   */
  public List<UrlSegment> getQuerySegments() {
    return querySegments;
  }

  /**
   * Get all segments.
   *
   * @return list of segments.
   */
  public List<UrlSegment> getSegments() {
    return segments;
  }

  /**
   * Get suffix of path.
   *
   * @return suffix
   */
  public String getSuffix() {
    return suffix;
  }
}

File: plugin/internal/commons/segment_group.go
package commons

import (
	"regexp"
	"strings"
)

var (
	pathPattern  = regexp.MustCompile(`\{[^/]+}`)
	queryPattern = regexp.MustCompile(`(?:(\w+)=)?([^/?&]+)&?`)
)

// SegmentGroup groups UrlSegment values captured from a URL.
type SegmentGroup struct {
	pathSegments  []UrlSegment
	suffix        string
	querySegments []UrlSegment
	segments      []UrlSegment
}

// Of parses a URL into a SegmentGroup.
func Of(url string) SegmentGroup {
	group := SegmentGroup{}
	if url == "" {
		return group
	}
	parts := strings.SplitN(url, "?", 2)
	path := parts[0]
	query := ""
	if len(parts) == 2 {
		query = parts[1]
	} else if strings.Contains(path, "=") {
		path = ""
		query = parts[0]
	}

	if path != "" {
		loc := pathPattern.FindAllStringIndex(path, -1)
		index := 0
		for _, match := range loc {
			value := path[match[0]:match[1]]
			group.pathSegments = append(group.pathSegments, NewUrlSegment(path[index:match[0]], "", value))
			index = match[1]
		}
		group.suffix = path[index:]
	}

	if query != "" {
		submatches := queryPattern.FindAllStringSubmatch(query, -1)
		for _, m := range submatches {
			key := ""
			if len(m) > 1 {
				key = m[1]
			}
			value := ""
			if len(m) > 2 {
				value = m[2]
			}
			group.querySegments = append(group.querySegments, NewUrlSegment("", key, value))
		}
	}

	group.segments = append(group.segments, group.pathSegments...)
	group.segments = append(group.segments, group.querySegments...)
	return group
}

// IsEmpty reports whether the group has no segments.
func (g SegmentGroup) IsEmpty() bool {
	return len(g.pathSegments) == 0 && len(g.querySegments) == 0
}

// PathSegments returns path segments.
func (g SegmentGroup) PathSegments() []UrlSegment {
	return g.pathSegments
}

// QuerySegments returns query segments.
func (g SegmentGroup) QuerySegments() []UrlSegment {
	return g.querySegments
}

// Segments returns all segments.
func (g SegmentGroup) Segments() []UrlSegment {
	return g.segments
}

// Suffix returns the path suffix.
func (g SegmentGroup) Suffix() string {
	return g.suffix
}
